File: java/src/main/java/S044.java
import com.google.common.math.DoubleMath;

import java.math.RoundingMode;

public class S044 {

    public static long solve() {
        long minDelta = Long.MAX_VALUE;
        for (int top = 2;; top++) {
            for (int bottom = top - 1; bottom >= 1; bottom--) {
                long pM = pentagonal(top);
                long pN = pentagonal(bottom);
                if (pM - pN >= minDelta) {
                    break;
                }
                if (isSpecial(pN, pM)) {
                    minDelta = Math.min(minDelta, pM - pN);
                }
            }
            if (pentagonal(top + 1) - pentagonal(top) > minDelta) {
                return minDelta;
            }
        }
    }

    // m > n
    private static boolean isSpecial(long pN, long pM) {
        return isPentagonal(pM - pN) && isPentagonal(pM + pN);
    }

    private static boolean isPentagonal(long p) {
        long candidate = DoubleMath.roundToLong((1 + Math.sqrt(1 + 24 * p)) / 6, RoundingMode.DOWN);
        return pentagonal(candidate) == p;
    }

    private static long pentagonal(long d) {
        return d * (3 * d - 1) / 2;
    }
}

File: cpp/S044.cpp
#include <iostream>
#include <climits>
#include "maths.cpp"

using namespace std;

static const int LIMIT = 100 * 1000;

inline bool is_pentagonal(int n) {
    int maybe_idx = (1 + sqrt(1 + 24 * n)) / 6;
    return pentagonal(maybe_idx) == n;
}

int main() {
    int pentagonals[LIMIT];
    for (int i = 1; i < LIMIT; i++) {
        pentagonals[i] = pentagonal(i);
    }
    int D = INT_MAX;
    for (int k = 2; k < LIMIT; k++) {
        if (pentagonals[k] > D) {
            break;
        }
        for (int j = 1; j < k; j++) {
            int pj = pentagonal(j), pk = pentagonal(k);
            if (is_pentagonal(pk - pj) && is_pentagonal(pk + pj)) {
                D = min(D, pk - pj);
            }
        }
    }
    cout << D << endl;
    return 0;
}
